From 7504be71cc80e3af12d1f532af8cbde07e78e9d9 Mon Sep 17 00:00:00 2001
From: Abhishek <apatgaonkar@habana.ai>
Date: Fri, 30 Aug 2024 08:34:32 +0300
Subject: [PATCH] [FSW-1175] fs1: recovery: get recovery image count

To load recovery images from each manifest image should
be counted. This patch parses all the descriptors and
counts recovery type images.

Change-Id: Iac5a4419bcb1107741bb3035404f495f3e08ea40
---

diff --git a/common/boot_desc.c b/common/boot_desc.c
index ec9f1eb..edea392 100644
--- a/common/boot_desc.c
+++ b/common/boot_desc.c
@@ -289,3 +289,31 @@
 
 	return BOOT_ROM_SUCCESS;
 }
+
+uint32_t boot_desc_get_recovery_img_count(enum hl_fw_manifest_type manifest_idx)
+{
+	struct hl_fw_img_desc *p_loc_img_desc = NULL;
+	uint32_t i, rec_img_count = 0;
+
+	if (manifest_idx >= HL_FW_SPI_MANIFEST_MAX_NUM)
+		return 0;
+
+	boot_desc_update_ptrs(manifest_idx);
+
+	if (boot_desc_validate_ptrs())
+		return 0;
+
+	for (i = 0; i < p_manifest_ctx->desc_info.count ; i++) {
+		p_loc_img_desc = boot_desc_img_desc_addr_calc(manifest_idx, i);
+
+		if (!p_loc_img_desc)
+			continue;
+
+		if (p_loc_img_desc->magic.type != HL_FW_IMG_RECOVERY)
+			continue;
+
+		rec_img_count++;
+	}
+
+	return rec_img_count;
+}
diff --git a/common/boot_desc.h b/common/boot_desc.h
index 5ad2c69..79334bc 100644
--- a/common/boot_desc.h
+++ b/common/boot_desc.h
@@ -36,6 +36,7 @@
 uint32_t boot_rom_verify_img_desc_address_range(struct boot_desc *desc,
 						enum boot_rom_progress img_sts_base,
 						enum hl_fw_img_type img);
+uint32_t boot_desc_get_recovery_img_count(enum hl_fw_manifest_type manifest_idx);
 
 #endif /* _BOOT_DESC_H_ */
 
diff --git a/common/boot_flow.c b/common/boot_flow.c
index d96cb84..324a5c1 100644
--- a/common/boot_flow.c
+++ b/common/boot_flow.c
@@ -214,3 +214,16 @@
 	/* Validate ppboot sha */
 	return true;
 }
+
+uint32_t boot_rom_boot_flow_get_recovery_img_count(enum boot_rom_manifests recovery_manifest)
+{
+	enum hl_fw_manifest_type manifest_idx = primary_manifest_idx;
+
+	if (recovery_manifest == BOOT_ROM_SECONDARY)
+		manifest_idx = (primary_manifest_idx + 1) % HL_FW_SPI_MANIFEST_MAX_NUM;
+
+	if (!boot_rom_boot_flow_is_manifest_valid(recovery_manifest))
+		return 0;
+
+	return boot_desc_get_recovery_img_count(manifest_idx);
+}
diff --git a/common/boot_flow.h b/common/boot_flow.h
index da5de0c..2efabad 100644
--- a/common/boot_flow.h
+++ b/common/boot_flow.h
@@ -14,5 +14,6 @@
 bool boot_rom_boot_flow_is_ppboot_valid(enum boot_rom_manifests rom_manifest_type);
 bool boot_rom_boot_flow_is_secondary_manifest_valid(void);
 void boot_rom_boot_flow_wait_for_dma_finish(void);
+uint32_t boot_rom_boot_flow_get_recovery_img_count(enum boot_rom_manifests recovery_manifest);
 
 #endif /* _BOOT_FLOW_H_ */
diff --git a/common/recovery.c b/common/recovery.c
index 8974f9e..8557032 100644
--- a/common/recovery.c
+++ b/common/recovery.c
@@ -30,8 +30,18 @@
 	boot_rom_status_update(PROGRESS_BOOT_ROM_RECOVERY_END);
 }
 
-static int boot_rom_execute_recovery_img(__unused enum boot_rom_manifests recovery_manifest)
+static int boot_rom_execute_recovery_img(enum boot_rom_manifests recovery_manifest)
 {
+	uint32_t rec_img_count = 0;
+	enum boot_rom_progress img_sts_base = PROGRESS_RECOVERY_IMG_IMG0;
+
+	boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_START);
+	rec_img_count = boot_rom_boot_flow_get_recovery_img_count(recovery_manifest);
+	if (!rec_img_count) {
+		boot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_NO_IMG_DESC);
+		return BOOT_ROM_IMAGE_NOT_FOUND;
+	}
+
 	return BOOT_ROM_SUCCESS;
 }
 
